src/D3MHub.sol

```diff
@@ -221,121 +219,123 @@ contract D3MHub {
+            vat.slip(ilk, address(_pool), -int256(amount));
+        }
+        emit Unwind(ilk, amount);
+    }
 
-        // Unwind amount is limited by how much:
-        // - max reduction desired
-        // - assets available
-        // - dai debt tracked in vat (CDP or free)
-        uint256 amount = _min(
-                            _min(
-                                supplyReduction,
-                                availableAssets
-                            ),
-                            daiDebt
-                        );
-        require(amount <= MAXINT256, "D3MHub/overflow");
+    function _fullUnwind(bytes32 ilk, ID3MPool _pool, address urn) internal {
+        uint256 amount = _pool.maxWithdraw();
+        if (amount > 0) {
+            uint256 toSlip = _min(vat.gem(ilk, urn), amount);
+            require(toSlip <= MAXINT256, "D3MHub/overflow");
 
-        // Determine the amount of fees to bring back
-        uint256 fees = 0;
-        if (assetBalance > daiDebt) {
-            unchecked {
-                fees = assetBalance - daiDebt;
-            }
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
 
-            if (amount + fees > availableAssets) {
-                unchecked {
-                    fees = availableAssets - amount;
-                }
-            }
+            vat.slip(ilk, urn, -int256(toSlip));
+            vat.move(address(this), vow, amount * RAY);
         }
+        emit Unwind(ilk, amount);
+    }
+
+    function _normal(bytes32 ilk, ID3MPool _pool, uint256 Art, uint256 lineWad) internal {
+        uint256 currentAssets = _pool.assetBalance(); // Should return DAI owned by D3MPool
 
-        if (amount == 0 && fees == 0) {
-            emit Unwind(ilk, 0, 0);
-            return;
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(_pool));
+        uint256 maxWithdraw = _pool.maxWithdraw();
+        if (currentAssets > ink) { // If fees were generated
+            uint256 fixInk = _min(
+                currentAssets - ink, // fees generated
+                ink <= lineWad // if previously CDP was under debt ceiling
+                    ? lineWad + maxWithdraw - ink // we can just go above debt ceiling + maxWithdraw
+                    : maxWithdraw // we can just go above current state + maxWithdraw
+            );
+            require(fixInk <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, address(_pool), int256(fixInk)); // Generate extra collateral
+            vat.frob(ilk, address(_pool), address(_pool), address(this), int256(fixInk), 0); // Lock it
+            ink += fixInk;
+            emit Fees(ilk, fixInk);
+        }
+        if (art < ink) { // If there was permissionless DAI paid or fees added as collateral
+            address _vow = vow;
+            uint256 fixArt = ink - art; // Amount of fees + permissionless DAI paid we will now transform to debt
+            delete art;
```

Comment:
Is this needed?

Reply:
Not really, I put it just to see it helped with stack too deep (that I used to get), but mainly to show that variable is discarded. As the `ink` and `Art` are updated, it felt a bit weird leaving the `art` non updated, but also didn't want to update it if not used anymore.

Reply:
Was not even aware of this. You can free up stack slots with `delete`...? 🤯

Reply:
I don't think you can, iirc it just assigns 0 to it.

Reply:
Yeah yeah, not sure if actually does something about the stack too deep. Probably not as Tal says.